### PR TITLE
Release event pointers faster as they go through the pipeline

### PR DESCRIPTION
**Pull request by faec**
## Proposed commit message

Make two changes to allow event data to be freed from memory faster:
- Free event pointers from the memory queue buffer when they are vended instead of when they're acknowledged. (The data will be preserved in the event batch structure until acknowledgment.)
- Free event pointers from the batch structure immediately after it is acknowledged instead of waiting for the batch to be freed naturally.

In benchmarks of Filebeat with saturated Filestream input going to an Elasticsearch output, this lowered average memory footprint by ~10%.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

File: libbeat/publisher/pipeline/ttl_batch.go [hunk 1]
```diff
@@ -93,10 +93,14 @@ func (b *ttlBatch) Events() []publisher.Event {
...
 func (b *ttlBatch) ACK() {
+	// Help the garbage collector clean up the event data a little faster
+	b.events = nil
 	b.done()
 }
 
 func (b *ttlBatch) Drop() {
+	// Help the garbage collector clean up the event data a little faster
+	b.events = nil
 	b.done()
```
Comment (ycombinator) [inline libbeat/publisher/pipeline/ttl_batch.go:104]: Would it be possible to add unit tests around this change?

**Review by ycombinator — COMMENTED** [review]
Changes LGTM.  Would it be possible to add some unit tests, basically ensuring the appropriate events fields are `nil`'d out after calls to the methods changed in this PR?

File: libbeat/publisher/pipeline/ttl_batch.go [hunk 2]
```diff
@@ -93,10 +93,14 @@ func (b *ttlBatch) Events() []publisher.Event {
 }
 
 func (b *ttlBatch) ACK() {
+	// Help the garbage collector clean up the event data a little faster
+	b.events = nil
```
Comment (cmacknz) [inline libbeat/publisher/pipeline/ttl_batch.go:97]: Should this (and the one below) be done after calling `b.done()` in case the implementation of `done` references the contents of `events` some time in the future?

Comment (faec) [reply libbeat/publisher/pipeline/ttl_batch.go:97]: The done callback is just to forward acknowledgments (and most of that is async, so changing the order wouldn't help), but it has no access to the contents of the batch

**Review by leehinman — APPROVED** [review]
LGTM

**Review by ycombinator — APPROVED** [review]
